In [3]:
import os
import pandas as pd
import numpy as np 
from sklearn.model_selection import train_test_split

In [4]:
PROJECT_DIR=r'C:\Users\Harsha\Desktop\DataScience\ML\Projects\flightSagemakerr'
DATA_DIR=r'data'

In [5]:
def get_data(file):
    file_name=f"{file}.csv"
    file_path=os.path.join(PROJECT_DIR,DATA_DIR,file_name)
    return pd.read_csv(file_path)

In [6]:
flights=get_data("flight_price")

In [7]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


- The dataset contains 10683 rows ans 11 columns
- Columns Route and Total_stops contain null values 

## 3.Premelinary analysis

### 3.1 Check DataTypes

In [32]:
flights.head(30)

Airline Date_of_Journey    Source Destination  \
0              IndiGo      24/03/2019  Banglore   New Delhi   
1           Air India       1/05/2019   Kolkata    Banglore   
2         Jet Airways       9/06/2019     Delhi      Cochin   
3              IndiGo      12/05/2019   Kolkata    Banglore   
4              IndiGo      01/03/2019  Banglore   New Delhi   
5            SpiceJet      24/06/2019   Kolkata    Banglore   
6         Jet Airways      12/03/2019  Banglore   New Delhi   
7         Jet Airways      01/03/2019  Banglore   New Delhi   
8         Jet Airways      12/03/2019  Banglore   New Delhi   
9   Multiple carriers      27/05/2019     Delhi      Cochin   
10          Air India       1/06/2019     Delhi      Cochin   
11             IndiGo      18/04/2019   Kolkata    Banglore   
12          Air India      24/06/2019   Chennai     Kolkata   
13        Jet Airways       9/05/2019   Kolkata    Banglore   
14             IndiGo      24/04/2019   Kolkata    Banglore   
15          Air India       3/03/2019     Delhi      Cochin   
16           SpiceJet      15/04/2019     Delhi      Cochin   
17        Jet Airways      12/06/2019     Delhi      Cochin   
18          Air India      12/06/2019     Delhi      Cochin   
19        Jet Airways      27/05/2019     Delhi      Cochin   
20              GoAir       6/03/2019     Delhi      Cochin   
21          Air India      21/03/2019  Banglore   New Delhi   
22             IndiGo       3/04/2019  Banglore       Delhi   
23             IndiGo       1/05/2019  Banglore       Delhi   
24        Jet Airways       6/05/2019   Kolkata    Banglore   
25        Jet Airways       9/06/2019     Delhi      Cochin   
26             IndiGo       1/06/2019     Delhi      Cochin   
27              GoAir      15/05/2019     Delhi      Cochin   
28            Vistara      18/06/2019  Banglore       Delhi   
29            Vistara      15/06/2019   Chennai     Kolkata   

                    Route Dep_Time  Arrival_Time Duration Total_Stops  \
0               BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1   CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2   DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3         CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4         BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
5               CCU → BLR    09:00         11:25   2h 25m    non-stop   
6         BLR → BOM → DEL    18:55  10:25 13 Mar  15h 30m      1 stop   
7         BLR → BOM → DEL    08:00  05:05 02 Mar   21h 5m      1 stop   
8         BLR → BOM → DEL    08:55  10:25 13 Mar  25h 30m      1 stop   
9         DEL → BOM → COK    11:25         19:15   7h 50m      1 stop   
10        DEL → BLR → COK    09:45         23:00  13h 15m      1 stop   
11              CCU → BLR    20:20         22:55   2h 35m    non-stop   
12              MAA → CCU    11:40         13:55   2h 15m    non-stop   
13        CCU → BOM → BLR    21:10  09:20 10 May  12h 10m      1 stop   
14              CCU → BLR    17:15         19:50   2h 35m    non-stop   
15  DEL → AMD → BOM → COK    16:40  19:15 04 Mar  26h 35m     2 stops   
16        DEL → PNQ → COK    08:45         13:15   4h 30m      1 stop   
17        DEL → BOM → COK    14:00  12:35 13 Jun  22h 35m      1 stop   
18  DEL → CCU → BOM → COK    20:15  19:15 13 Jun      23h     2 stops   
19        DEL → BOM → COK    16:00  12:35 28 May  20h 35m      1 stop   
20        DEL → BOM → COK    14:10         19:20   5h 10m      1 stop   
21        BLR → COK → DEL    22:00  13:20 19 Mar  15h 20m      1 stop   
22              BLR → DEL    04:00         06:50   2h 50m    non-stop   
23              BLR → DEL    18:55         21:50   2h 55m    non-stop   
24        CCU → BOM → BLR    18:55  08:15 07 May  13h 20m      1 stop   
25  DEL → IDR → BOM → COK    21:25  12:35 10 Jun  15h 10m     2 stops   
26        DEL → LKO → COK    21:50  03:35 02 Jun   5h 45m      1 stop   
27        DEL → 

In [9]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [10]:
value=flights.Date_of_Journey.iloc[6]
value

'12/03/2019'

In [11]:
value=flights.Dep_Time.iloc[6]
value

'18:55'

In [12]:
value=flights.Arrival_Time.iloc[6]
value

'10:25 13 Mar'

### 3.2 Check for Duplicates

In [13]:
flights.duplicated().sum()

220

In [14]:
(
    flights
    .loc[flights.duplicated(keep=False)]
    .sort_values(["Airline","Date_of_Journey","Source","Destination"])
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

### 3.3 Observations

- The type of "Date_of_Journey","Dep_Time" and "Arrival_Time" should be datetime
- The type of duration is mixed so it should be numeric
- There are 220 duplicates These should be removed

## 4 Detailed Analysis

In [15]:
flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

- some of the entries have inconsistence and inaccurtate  values


### Date Of Journey

In [16]:
flights.Date_of_Journey

0        24/03/2019
1         1/05/2019
2         9/06/2019
3        12/05/2019
4        01/03/2019
            ...    
10678     9/04/2019
10679    27/04/2019
10680    27/04/2019
10681    01/03/2019
10682     9/05/2019
Name: Date_of_Journey, Length: 10683, dtype: object

In [17]:
pd.to_datetime(flights.Date_of_Journey,dayfirst=True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

### Source  and destination

In [18]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

In [19]:
flights.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

##### DEP Time

In [20]:
flights.Dep_Time

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

In [21]:
(
flights
.Duration
.loc[lambda ser:~ser.str.contains("h")])

6474    5m
Name: Duration, dtype: object

In [22]:
flights.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

- This observation 6474 has duration of 5 minutes.This is clearly wrong observation

In [23]:
(flights
 .Duration
 .drop(index=6474)
 .str.split(" ",expand=True)
 .set_axis(['hour','minute'],axis=1)
 .assign(
     hour=lambda df_:(
        df_
        .hour
        .str.replace("h","")
         .astype(int)
         .mul(60)
        ),
     minute=lambda df_:(
         df_
         .minute
         .str.replace("m","")
         .fillna("0")
         .astype(int)
     )
     )
 .sum(axis=1)
 .rename("duration_minutes")
 .to_frame()
 .join(flights.Duration.drop(index=6474))
)

duration_minutes Duration
0                   170   2h 50m
1                   445   7h 25m
2                  1140      19h
3                   325   5h 25m
4                   285   4h 45m
...                 ...      ...
10678               150   2h 30m
10679               155   2h 35m
10680               180       3h
10681               160   2h 40m
10682               500   8h 20m

[10682 rows x 2 columns]

#### Total Stops

In [24]:
flights.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [25]:
(
    flights
    .Total_Stops
    .replace("non-stop","0")
    .replace(" stops?","",regex=True)
    .pipe(lambda ser: pd.to_numeric(ser))
    .fillna(method="ffill")
    .isnull().sum()

)

0

#### Additional information

In [26]:
flights.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

## 5.Cleaning Operations

In [27]:
def convert_to_minutes(ser):
    return(
        ser
        .str.split(" ",expand=True)
     .set_axis(['hour','minute'],axis=1)
     .assign(
         hour=lambda df_:(
            df_
            .hour
            .str.replace("h","")
             .astype(int)
             .mul(60)
            ),
         minute=lambda df_:(
             df_
             .minute
             .str.replace("m","")
             .fillna("0")
             .astype(int)
             )
         )
     .sum(axis=1)
    )

In [30]:
def clean_data(df):
    return(
        df
        .drop(index=[6474])
        .assign(**{
            col:df[col].str.strip()
            for col in df.select_dtypes(include="O").columns   
        })
        .rename(columns=str.lower)
        .assign(
            airline=lambda df_: (
                df_
                .airline
                .str.replace("Premium economy","")
                .str.replace("Business","")
                .str.title()
            ),
            date_of_journey=lambda df_: pd.to_datetime(df_.date_of_journey,dayfirst=True),
            dep_time=lambda df_: pd.to_datetime(df_.dep_time).dt.time,
            arrival_time=lambda df_: pd.to_datetime(df_.arrival_time).dt.time,
            duration=lambda df_: df_.duration.pipe(convert_to_minutes),
            total_stops=lambda df_:(
                df_
                .total_stops
                .replace("non-stop","0")
                .replace(" stops?","",regex=True)
                .fillna(method="ffill")
                .pipe(lambda ser: pd.to_numeric(ser))
            ),
            aditional_info=lambda df_:df_.additional_info.replace("No info","No Info")
        )
        .drop(columns="route"),
        
    )

In [35]:
flights_cleaned=clean_data(flights)


C:\Users\Harsha\AppData\Local\Temp\ipykernel_8744\1454190106.py:19: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time=lambda df_: pd.to_datetime(df_.dep_time).dt.time,
C:\Users\Harsha\AppData\Local\Temp\ipykernel_8744\1454190106.py:20: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time=lambda df_: pd.to_datetime(df_.arrival_time).dt.time,


In [39]:
flights_cleaned

(           airline date_of_journey    source destination  dep_time  \
 0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
 1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
 2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
 3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
 4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
 ...            ...             ...       ...         ...       ...   
 10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
 10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
 10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
 10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
 10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   
 
       arrival_time  duration  total_stops additional_info  price  \
 0         01:10:00       170            0         No info   3897   
 1      

## 6.Split The data

## 7.Export the Subset